### ISSUE TITLE
[WARNING] Unknown keyword meta:[enum|deprecated] - you should define your own Meta Schema

### ISSUE THREAD

**Issue by garydgregory** [CycloneDX/cyclonedx-maven-plugin#564]
Hi All,

What am I supposed to do with these warnings:

```
[INFO] --- cyclonedx:2.9.0:makeAggregateBom (build-sbom-cyclonedx) @ commons-cli ---
[INFO] CycloneDX: Resolving Dependencies
[INFO] CycloneDX: Creating BOM version 1.6 with 0 component(s)
[INFO] CycloneDX: Writing and validating BOM (XML): /Users/garydgregory/git/commons-cli/target/commons-cli-1.9.1-SNAPSHOT-bom.xml
[INFO]            attaching as commons-cli-1.9.1-SNAPSHOT-cyclonedx.xml
[INFO] CycloneDX: Writing and validating BOM (JSON): /Users/garydgregory/git/commons-cli/target/commons-cli-1.9.1-SNAPSHOT-bom.json
[WARNING] Unknown keyword meta:enum - you should define your own Meta Schema. If the keyword is irrelevant for validation, just use a NonValidationKeyword or if it should generate annotations AnnotationKeyword
[WARNING] Unknown keyword deprecated - you should define your own Meta Schema. If the keyword is irrelevant for validation, just use a NonValidationKeyword or if it should generate annotations AnnotationKeyword
[INFO]            attaching as commons-cli-1.9.1-SNAPSHOT-cyclonedx.json
?
```

Reproducer:

```
git clone https://gitbox.apache.org/repos/asf/commons-cli
cd commons-cli
mvn clean verify -Dcommons.cyclonedx.version=2.9.0 -DskipTest
```

TY

**Comment by hboutemy** [comment]
thanks for the report: it seems #305 is going worse, with root cause https://github.com/CycloneDX/cyclonedx-core-java/issues/280

notice it seems the precise keyword issue from #305 seems to have been fixed by cyclonedx-core-java, but new similar issues addeed

I confess I opened https://github.com/CycloneDX/cyclonedx-core-java/issues/280 but I did not try to help fixing: this is definitively the project requiring additional love...

**Comment by patpatpat123** [comment]
Hello team, I am facing the same issue.

Could you please help on this?

Thank you

**Comment by hboutemy** [comment]
> Could you please help on this?

yes: I added the "help wanted" label to clarify what's needed = someone to work on it, it's OSS

**Comment by garydgregory** [comment]
Arg, still happening, I guess we're all too busy to help.

**Comment by hboutemy** [comment]
perhaps nowadays, a good AI prompt could help us get a global transversal analysis on the topic?

**Comment by patpatpat123** [comment]
Here is using GPT 5.5 xHIGH:

• Assumptions

  - The reported run uses cyclonedx-maven-plugin:2.9.0 with default schemaVersion=1.6 and JSON output enabled.
  - The warnings are emitted during JSON validation, not during BOM generation.
  - No claim is made here about newer upstream releases unless verified from local artifacts.

  Root Cause
  The warnings are caused by CycloneDX 1.6 JSON schema validation, not by the generated BOM content.

  Evidence:

  - Plugin 2.9.0 depends on cyclonedx-core-java:9.0.5; the current checkout also declares that dependency in pom.xml:104.
  - schemaVersion defaults to 1.6 in src/main/java/org/cyclonedx/maven/BaseCycloneDxMojo.java:85.
  - JSON output is generated and validated via new JsonParser() in src/main/java/org/cyclonedx/maven/BaseCycloneDxMojo.java:449.
  - The plugin writes the BOM, then calls bomParser.isValid(bomFile, schemaVersion()) in src/main/java/org/cyclonedx/maven/BaseCycloneDxMojo.java:457.
  - cyclonedx-core-java:9.0.5 depends on com.networknt:json-schema-validator:1.5.1 in /Users/patrickpok/.m2/repository/org/cyclonedx/cyclonedx-core-java/9.0.5/cyclonedx-core-java-9.0.5.pom:189.
  - The bundled bom-1.6.schema.json contains 32 occurrences of "meta:enum" and 4 occurrences of "deprecated".
  - NetworkNT’s UnknownKeywordFactory contains the exact warning text and logs it at WARN when schema keywords are not registered.

  So the concrete cause is:

  cyclonedx-core-java:9.0.5 loads the CycloneDX 1.6 JSON schema with NetworkNT’s standard draft-07 meta-schema, but CycloneDX 1.6’s schema includes annotation/documentation keywords meta:enum and deprecated
  that are not registered in that meta-schema. NetworkNT therefore warns once for each unknown keyword name.

  Impact
  The BOM is not invalid because of these warnings. NetworkNT treats the unknown keyword as an annotation after logging the warning, and the plugin continues to attach the JSON artifact.

  The 0 component(s) in the sample log is unrelated. These warnings come from loading the schema, not from validating a particular component list.

  Fix
  The correct fix belongs in cyclonedx-core-java, not in generated BOM content.

  Register the CycloneDX schema annotation keywords before creating the JSON schema validator, for example in CycloneDxSchema.getJsonSchema(...):

  JsonMetaSchema cycloneDxV7MetaSchema = JsonMetaSchema.builder(JsonMetaSchema.getV7())
      .keyword(new AnnotationKeyword("meta:enum"))
      .keyword(new AnnotationKeyword("deprecated"))
      .build();

  JsonSchemaFactory factory = JsonSchemaFactory
      .builder(JsonSchemaFactory.getInstance(SpecVersionDetector.detect(schemaNode)))
      .jsonMapper(mapper)
      .metaSchema(cycloneDxV7MetaSchema)
      .schemaMappers(s -> s.add(offlineSchemaMapper))
      .build();

  Then update this Maven plugin to depend on the fixed cyclonedx-core-java version.

  Do Not Fix By

  - Changing generated BOM output. The warning is from schema parsing, not BOM fields.
  - Removing meta:enum or deprecated from the bundled CycloneDX schema. Those are schema metadata.
  - Suppressing all com.networknt.schema.UnknownKeywordFactory warnings as the primary fix. That hides the symptom and may mask real schema issues later.

  Regression Test
  Add an integration test that runs JSON generation with default schemaVersion=1.6 and asserts the build log does not contain:

  Unknown keyword meta:enum
  Unknown keyword deprecated

  Also keep asserting that JSON validation succeeds.

**Comment by patpatpat123** [comment]
Here using Opus 4.8 MAX

⏺ Assumptions

  1. I analyzed the checked-out repository (cyclonedx-maven-plugin 2.9.2-SNAPSHOT). The log you pasted is from released 2.9.0, but the relevant code path and dependency chain are materially identical; the root
  cause is version-independent.
  2. The -bom.json being written is CycloneDX schema 1.6 — the plugin's default (BaseCycloneDxMojo.java:85, defaultValue = "1.6") and consistent with the log line "Creating BOM version 1.6".
  3. The two [WARNING] lines are SLF4J log output routed through Maven's maven-slf4j-provider (slf4j-simple fork), at WARN level.

  Root Cause

  The warnings are emitted by the com.networknt:json-schema-validator library (pulled in transitively by cyclonedx-core-java), not by the plugin's own code. They fire because the CycloneDX 1.6 JSON schema 
  declares JSON Schema draft-07 but contains two keywords that draft-07 does not define, and cyclonedx-core-java does not register those keywords with the validator.

  Evidence chain

  1. Plugin → core-java → validator versions
    - pom.xml:106-109 → org.cyclonedx:cyclonedx-core-java:9.0.5.
    - cyclonedx-core-java-9.0.5 POM pins com.networknt:json-schema-validator:1.5.1.
  2. Where validation is triggered
    - BaseCycloneDxMojo.saveBomToFile calls bomParser.isValid(bomFile, schemaVersion()) (BaseCycloneDxMojo.java:463) — this is the "Writing and validating BOM (JSON)" step.
    - JsonParser.validate(...) → getJsonSchema(schemaVersion, mapper).validate(...) (in cyclonedx-core-java's JsonParser.java).
  3. How the validator is configured (the actual defect site)
    - CycloneDxSchema.getJsonSchema(...) builds the factory via JsonSchemaFactory.builder(JsonSchemaFactory.getInstance(SpecVersionDetector.detect(schemaNode))). Because bom-1.6.schema.json declares "$schema":
  "http://json-schema.org/draft-07/schema#", the detected spec is draft-07 (V7).
    - The SchemaValidatorsConfig it passes sets only setPreloadJsonSchema(false). It does not register the proprietary keywords and does not install a custom (silent) unknown-keyword handler.
  4. The offending keywords exist only in the 1.6 schema
    - bom-1.6.schema.json: meta:enum appears 32×, "deprecated" appears 4×.
    - bom-1.2 … 1.5 schemas: zero occurrences of either. This is exactly why the warning appears only at the (default) 1.6 version.
    - meta:enum is a CycloneDX-proprietary annotation keyword (no JSON Schema draft defines it). deprecated was introduced in draft 2019-09, so it is not part of the draft-07 vocabulary the file declares.
  5. Why the validator logs (and that it is harmless)
    - In networknt's JsonMetaSchema.java:489-491, any keyword not in the meta-schema's known set falls back to UnknownKeywordFactory.getInstance().getKeyword(...) (no custom factory was configured).
    - UnknownKeywordFactory.java:35-40 logs, once per distinct keyword (cached in a ConcurrentHashMap, on a process-wide singleton), exactly the message you see, then returns new AnnotationKeyword(keyword).
    - Version7.java / ValidatorTypeCode.java confirm draft-07's keyword set excludes both deprecated and meta:enum (only readOnly/writeOnly exist as V7 annotation keywords).
    - Because the fallback is an AnnotationKeyword (non-asserting), validation still passes — your log proceeds to "attaching as …cyclonedx.json" with no failure. The warnings are pure noise, not a validation
  error.

  Net: harmless WARN logs caused by a schema/validator mismatch — the bundled 1.6 schema uses meta:enum and deprecated, which the draft-07-configured networknt validator doesn't recognize, and
  cyclonedx-core-java never declares them as known/non-validating keywords.

  How to Fix

  The defect is in cyclonedx-core-java (CycloneDxSchema.getJsonSchema), not in this plugin. There are two distinct levels of remedy.

  A. Consumer-side mitigation (no code change; only suppresses the noise)

  Anyone running the plugin (e.g. the commons-cli build) can silence it by raising the log threshold for the emitting class, since it goes through Maven's slf4j-simple provider:

  mvn … -Dorg.slf4j.simpleLogger.log.com.networknt.schema.UnknownKeywordFactory=error

  (or target the package com.networknt.schema). Can be made persistent via .mvn/maven.config, MAVEN_OPTS, or MAVEN_ARGS. This hides the symptom only; it does not change that the keywords are unrecognized.

  B. True fix — in cyclonedx-core-java

  In CycloneDxSchema.getJsonSchema(...), register the extra keywords against a derived draft-07 meta-schema so the validator treats them as known. The networknt 1.5.1 API supports this directly
  (JsonMetaSchema.builder(iri, blueprint) at line 391; .keyword(...) at 155; getV7() at 354):

  JsonMetaSchema base = JsonMetaSchema.getV7();
  JsonMetaSchema custom = JsonMetaSchema.builder(base.getIri(), base)
          .keyword(new NonValidationKeyword("meta:enum"))
          .keyword(new AnnotationKeyword("deprecated"))
          .build();

  JsonSchemaFactory factory = JsonSchemaFactory
          .builder(JsonSchemaFactory.getInstance(SpecVersion.VersionFlag.V7))
          .metaSchema(custom)
          .jsonMapper(mapper)
          .schemaMappers(s -> s.add(offlineSchemaMapper))
          .build();

  Alternative within the same call: set a silent unknown-keyword handler via JsonMetaSchema.Builder.unknownKeywordFactory(...) (line 122) returning NonValidationKeyword without logging. Both eliminate the WARN
  at the source while preserving identical validation results (the keywords are non-asserting either way).

  C. What this plugin repository can realistically do

  This repo does not own CycloneDxSchema, so it cannot fix the warning in its own source. Its only options are:
  - Upgrade the cyclonedx-core-java dependency (pom.xml:108) to a release that incorporates fix B upstream — the correct long-term action, contingent on such a release existing (I have not verified one in the
  local repository; do not assume a specific fixed version without checking).
  - Ship mitigation A as documented guidance, or programmatically lower the com.networknt.schema logger level before invoking validation — a hack with global logging side effects, not recommended.
  
  Bumping json-schema-validator alone will not help: the UnknownKeywordFactory warn-on-unknown behavior is the library's default and is present in 1.5.1; the keywords must be registered (B) or the log
  suppressed (A).

**Comment by garydgregory** [comment]
FWIW, according to https://json-schema.org/specification version "draft-07" is two versions behind the current version.

**Comment by patbaumgartner** [comment]
Still reproducing with **cyclonedx-maven-plugin 2.9.1** (which pins **cyclonedx-core-java 9.0.5** → **networknt json-schema-validator 1.5.1**), default `schemaVersion=1.6`, JSON output:

```
[WARNING] Unknown keyword meta:enum - you should define your own Meta Schema. If the keyword is irrelevant for validation, just use a NonValidationKeyword or if it should generate annotations AnnotationKeyword
[WARNING] Unknown keyword deprecated - you should define your own Meta Schema. If the keyword is irrelevant for validation, just use a NonValidationKeyword or if it should generate annotations AnnotationKeyword
```

Root cause is as already described in this thread: `bom-1.6.schema.json` declares draft-07 but uses the proprietary `meta:enum` (and the `deprecated`) keyword, which the validator doesn't know, so networknt's `UnknownKeywordFactory` logs once per keyword and falls back to a non-asserting annotation. The BOM is still valid — it's pure log noise.

One thing worth flagging for anyone picking this up: the fix site `CycloneDxSchema.getJsonSchema(...)` on `master` has **already been migrated to the networknt 2.x API** (`json-schema-validator 2.0.1`, `SchemaRegistry` / `SchemaRegistryConfig`), so the `JsonMetaSchema.builder().keyword(...)` approach proposed earlier (written against networknt 1.5.x) no longer applies as-is. On `master` the two keywords would need to be registered — or a silent unknown-keyword handler installed — via the new `SchemaRegistry` builder.

Consumer-side mitigation that works today without disabling validation (raises the log threshold only for the emitting class, since the plugin logs through Maven's slf4j-simple provider):

```
mvn ... -Dorg.slf4j.simpleLogger.log.com.networknt.schema.UnknownKeywordFactory=error
```

(persistable via `.mvn/maven.config`, `MAVEN_OPTS`, or `MAVEN_ARGS`).

Happy to open a PR against `master` that registers `meta:enum` and `deprecated` as non-validating keywords in `getJsonSchema`, plus a regression test asserting the build log no longer contains `Unknown keyword meta:enum` / `Unknown keyword deprecated` while JSON validation still passes — if a maintainer is open to it.

**Comment by garydgregory** [comment]
I'm not a maintainer here but I'd like to see these warnings go away by default when this trickles down to the maven plug in so our Apache Commons builds output looks clean.

**Comment by hboutemy** [comment]
thank you @patbaumgartner , great fix in cyclonedx-core-java: now I know why I'll update the dependency once it is released :)

**Comment by VinodAnandan** [comment]
@hboutemy  The fixed version of cyclonedx-core-java v13.0.0 was released on 16 July. Thank you @patbaumgartner for the fix 

https://github.com/CycloneDX/cyclonedx-core-java/releases/tag/cyclonedx-core-java-13.0.0

**Comment by ThomasThelen** [comment]
Just wanted to say a big thank you for this fix, I can confirm that using the version from @VinodAnandan cleared up the warnings on our dependent project!
